src/main/java/aztech/modern_industrialization/machines/multiblocks/MultiblockMachineBlockEntity.java

```diff
@@ -45,7 +47,61 @@ public boolean isShapeValid() {
         return shapeValid.shapeValid;
     }
 
-    public abstract void unlink();
+    public ShapeMatcher getShapeMatcher() {
+        return shapeMatcher;
+    }
+
+    public ShapeMatcher createShapeMatcher() {
+        return new ShapeMatcher(level, worldPosition, orientation.facingDirection, getActiveShape());
+    }
+
+    protected void onLink() {
+    }
+
+    protected void onUnlink() {
+    }
+
+    protected void onRematch() {
+    }
+
+    protected void onMatchSuccessful() {
+    }
+
+    protected void onMatchFailure() {
+    }
+
+    protected void link() {
+        if (shapeMatcher == null) {
+            shapeMatcher = createShapeMatcher();
+            shapeMatcher.registerListeners(level);
+            onLink();
+        }
+        if (shapeMatcher.needsRematch()) {
+            shapeValid.shapeValid = false;
+            shapeMatcher.rematch(level);
+            onRematch();
+
+            if (shapeMatcher.isMatchSuccessful()) {
+                onMatchSuccessful();
+                shapeValid.shapeValid = true;
+            } else {
+                onMatchFailure();
```

Comment:
Current MI doesn't need this. Do you?

Reply:
Yeah, I use it for updating the network for a tesla tower when the multiblock forms or fails to form

Reply:
I prefer a reset in `onRematch()` than separate handling of `onMatchFailure()`.